File: DSA/src/Questions/Graph/ParallelCourses2.java
package Questions.Graph;

/*
https://leetcode.com/problems/parallel-courses-ii/description/
 You are given an integer n, which indicates that there are n courses labeled from 1 to n. You are also given an array relations where relations[i] = [prevCoursei, nextCoursei], representing a prerequisite relationship between course prevCoursei and course nextCoursei: course prevCoursei has to be taken before course nextCoursei. Also, you are given the integer k.

In one semester, you can take at most k courses as long as you have taken all the prerequisites in the previous semesters for the courses you are taking.

Return the minimum number of semesters needed to take all courses. The testcases will be generated such that it is possible to take every course.

 

Example 1:


Input: n = 4, relations = [[2,1],[3,1],[1,4]], k = 2
Output: 3
Explanation: The figure above represents the given graph.
In the first semester, you can take courses 2 and 3.
In the second semester, you can take course 1.
In the third semester, you can take course 4.
Example 2:


Input: n = 5, relations = [[2,1],[3,1],[4,1],[1,5]], k = 2
Output: 4
Explanation: The figure above represents the given graph.
In the first semester, you can only take courses 2 and 3 since you cannot take more than two per semester.
In the second semester, you can take course 4.
In the third semester, you can take course 1.
In the fourth semester, you can take course 5.
 

Constraints:

1 <= n <= 15
1 <= k <= n
0 <= relations.length <= n * (n-1) / 2
relations[i].length == 2
1 <= prevCoursei, nextCoursei <= n
prevCoursei != nextCoursei
All the pairs [prevCoursei, nextCoursei] are unique.
The given graph is a directed acyclic graph.
 */
public class ParallelCourses2 {
    public int minNumberOfSemesters(int n, int[][] relations, int k) {

        // total courses mask
        int course_mask = (1 << n) - 1;

        this.dp = new Integer[course_mask + 1];

        return helper(n, relations, course_mask, k);
    }

    Integer dp[];

    int helper(int n, int[][] relations, int course_mask, int k) {

        // if all courses are visited
        if (course_mask == 0)
            return 0;

        if (dp[course_mask] != null)
            return dp[course_mask];

        // calculating indegree
        int indeg[] = new int[n];
        for (int rel[] : relations) {
            int u = rel[0] - 1;
            int v = rel[1] - 1;
            if (isBitOn(course_mask, u))
                indeg[v]++;
        }

        // mask for total available course to take this semester
        int avail_course_mask = 0;
        for (int i = 0; i < indeg.length; i++) {
            if (indeg[i] == 0 && isBitOn(course_mask, i))
                avail_course_mask = setKthBit(avail_course_mask, i);
        }

        // total available course in this semester
        int total_available_course_this_sem = countSetBits(avail_course_mask);

        if (total_available_course_this_sem <= k) {
            int newMask = (course_mask ^ avail_course_mask);
            return dp[course_mask] = 1 + helper(n, relations, newMask, k);
        }

        // Generating all permutation of avail_course_mask in which K bit is ON.
        int min = n;
        for (int subs = avail_course_mask; subs > 0; subs = (subs - 1) & avail_course_mask) {
            if (countSetBits(subs) == k) {
                int newMask = (course_mask ^ subs);
                int temp = 1 + helper(n, relations, newMask, k);
                min = Math.min(min, temp);
            }
        }

        return dp[course_mask] = min;
    }

    boolean isBitOn(int num, int k) {
        return (num & (1 << k)) > 0;
    }

    int setKthBit(int num, int k) {
        return (num ^ (1 << k));
    }

    int countSetBits(int num) {
        int count = 0;
        while (num != 0) {
            num &= (num - 1);
            count++;
        }
        return count;
    }

   public static void main(String[] args) {
        int n = 4;
        int[][] relations = { { 2, 1 }, { 3, 1 }, { 1, 4 } };
        int k = 2;
        ParallelCourses2 parallelCourses2 = new ParallelCourses2();
        System.out.println(parallelCourses2.minNumberOfSemesters(n, relations, k));
   }
}

File: DSA/Python/Questions/Graph/TopologicalSort/parallel_courses2.py
"""
# Question: Parallel Courses II
# Link: https://leetcode.com/problems/parallel-courses-ii/

# Problem Statement:
# You are given an integer n, which indicates that there are n courses labeled from 1 to n.
# You are also given an array relations where relations[i] = [prevCoursei, nextCoursei],
# representing a prerequisite relationship between courses: course prevCoursei has to be
# taken before course nextCoursei. Also, you are given the integer k.
# Return the minimum number of semesters needed to take all courses. If there is no way
# to take all the courses, return -1.

# Example:
# Input: n = 4, dependencies = [[2,1],[3,1],[1,4]], k = 2
# Output: 3
# Explanation: The figure above represents the given graph.
# In the first semester, you can take courses 2 and 3.
# In the second semester, you can take course 1.
# In the third semester, you can take course 4.
"""

from typing import List
from collections import defaultdict, deque

# 1. BFS with State Compression
"""
Algorithm:
1. Build adjacency list and calculate in-degrees
2. Use BFS to process courses level by level
3. For each level, try all possible combinations of k courses
4. Track taken courses using bitmask
5. Return minimum number of semesters needed

Time Complexity: O(2^n * n)
Space Complexity: O(2^n)
"""


def min_number_of_semesters(n: int, relations: List[List[int]], k: int) -> int:
    # Build prerequisites graph
    prereq = defaultdict(list)
    in_degree = [0] * (n + 1)

    for prev, next in relations:
        prereq[prev].append(next)
        in_degree[next] += 1

    # Initialize queue with courses having no prerequisites
    queue = deque()
    for i in range(1, n + 1):
        if in_degree[i] == 0:
            queue.append(i)

    semesters = 0
    courses_taken = 0

    while queue:
        size = len(queue)
        if size > k:
            # Take k courses with maximum outgoing edges
            courses = list(queue)
            courses.sort(key=lambda x: len(prereq[x]), reverse=True)
            courses = courses[:k]
            queue = deque(courses)
            size = k

        # Process current semester
        for _ in range(size):
            course = queue.popleft()
            courses_taken += 1

            # Update prerequisites for dependent courses
            for next_course in prereq[course]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    queue.append(next_course)

        semesters += 1

    return semesters if courses_taken == n else -1


# 2. Dynamic Programming with Bitmask
"""
Algorithm:
1. Use bitmask to represent taken courses
2. For each state, try all possible combinations of k courses
3. Memoize results for each state
4. Check prerequisites before taking courses
5. Return minimum semesters needed

Time Complexity: O(2^n * nCk)
Space Complexity: O(2^n)
"""


def min_number_of_semesters_dp(n: int, relations: List[List[int]], k: int) -> int:
    # Create prerequisite mask for each course
    prereq = [0] * n
    for prev, next in relations:
        prereq[next - 1] |= 1 << (prev - 1)

    dp = [-1] * (1 << n)
    dp[0] = 0

    def solve(mask: int) -> int:
        if dp[mask] != -1:
            return dp[mask]

        # Find available courses
        available = 0
        for i in range(n):
            if not (mask & (1 << i)) and (prereq[i] & mask) == prereq[i]:
                available |= 1 << i

        if not available:
            return float("inf")

        # Try all combinations of k courses
        result = float("inf")
        curr = available
        while curr:
            if bin(curr).count("1") <= k:
                result = min(result, 1 + solve(mask | curr))
            curr = (curr - 1) & available

        dp[mask] = result
        return result

    result = solve(0)
    return result if result != float("inf") else -1


def main():
    test_cases = [
        {"n": 4, "relations": [[2, 1], [3, 1], [1, 4]], "k": 2, "expected": 3},
        {"n": 5, "relations": [[2, 1], [3, 1], [4, 1], [1, 5]], "k": 2, "expected": 4},
        {"n": 3, "relations": [[1, 2], [2, 3], [3, 1]], "k": 1, "expected": -1},
    ]

    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"n: {test['n']}")
        print(f"Relations: {test['relations']}")
        print(f"k: {test['k']}")
        print(f"Expected: {test['expected']}")
        print(
            f"BFS Solution: {min_number_of_semesters(test['n'], test['relations'], test['k'])}"
        )
        print(
            f"DP Solution: {min_number_of_semesters_dp(test['n'], test['relations'], test['k'])}"
        )


if __name__ == "__main__":
    main()
